politeiad/backendv2/tstorebe/tstorebe.go

```diff
@@ -0,0 +1,1251 @@
+
+	// Update inventory cache
+	switch status {
+	case backend.StatusPublic:
+		// The state is updated to vetted when a record is made public
+		t.inventoryMoveToVetted(token, status)
+	default:
+		t.inventoryUpdate(r.RecordMetadata.State, token, status)
+	}
+
+	// Return updated record
+	r, err = t.tstore.RecordLatest(treeID)
+	if err != nil {
+		return nil, fmt.Errorf("RecordLatest %v: %v", treeID, err)
+	}
+
+	return r, nil
+}
+
+// RecordExists returns whether a record exists.
+//
+// This method relies on the the tstore tree exists call. It's possible for a
+// tree to exist that does not correspond to a record in the rare case that a
+// tree was created but an unexpected error, such as a network error, was
+// encoutered prior to the record being saved to the tree. We ignore this edge
+// case because:
+//
+// 1. A user has no way to obtain this token unless the trillian instance has
+//    been opened to the public.
+//
+// 2. Even if they have the token they cannot do anything with it. Any attempt
+//  	to read from the tree or write to the tree will return a RecordNotFound
+//    error.
+//
+// Pulling the leaves from the tree to see if a record has been saved to the
+// tree adds a large amount of overhead to this call that should be very light
+// weight. Its for this reason that we rely on the tree exists call despite the
+// edge case.
+//
+// This function satisfies the Backend interface.
+func (t *tstoreBackend) RecordExists(token []byte) bool {
+	log.Tracef("RecordExists: %x", token)
+
+	// Read methods are allowed to use short tokens. Lookup the full
+	// length token.
+	var err error
+	token, err = t.fullLengthToken(token)
+	if err != nil {
+		return false
+	}
+
+	treeID := treeIDFromToken(token)
+	return t.tstore.TreeExists(treeID)
+}
+
+// RecordTimestamps returns the timestamps for a record. If no version is provided
+// then timestamps for the most recent version will be returned.
+//
+// This function satisfies the Backend interface.
+func (t *tstoreBackend) RecordTimestamps(token []byte, version uint32) (*backend.RecordTimestamps, error) {
+	log.Tracef("RecordTimestamps: %x %v", token, version)
+
+	// Read methods are allowed to use short tokens. Lookup the full
+	// length token.
+	var err error
+	token, err = t.fullLengthToken(token)
+	if err != nil {
+		return nil, err
+	}
+
+	treeID := treeIDFromToken(token)
+	return t.tstore.RecordTimestamps(treeID, version, token)
+}
+
+// Records retreives a batch of records. Individual record errors are not
+// returned. If the record was not found then it will not be included in the
+// returned map.
+//
+// This function satisfies the Backend interface.
+func (t *tstoreBackend) Records(reqs []backend.RecordRequest) (map[string]backend.Record, error) {
+	log.Tracef("Records: %v reqs", len(reqs))
+
+	records := make(map[string]backend.Record, len(reqs)) // [token]Record
+	for _, v := range reqs {
+		// Read methods are allowed to use short tokens. Lookup the full
```

Comment:
The more I see this the more inclined I am that token length handling needs to be pushed one layer down. We need to have this debate when we are going to get rid of the various memory and file caches.

Reply:
I don't disagree.  This is left over from when a record token didn't map directly to a tree ID and the tstorebe had to lookup the correct tree ID.  It makes sense to move it one layer down.

Reply:
This has been fixed.  The short token cache and token length handling has been moved from the `tstorebe` to the `tstore` instance.

Reply:
Better. Next step is killing this map.